Integrate DDP implementation

buianhhuy96 opened the pull request:
- Add new method to write json config for ddp service
- Add command to restart ddp service whenever the config changes

jsvapiav commented on utils/fmoconfig.py:
```diff
@@ -156,6 +161,25 @@ def set_vmddp_config(self, vmname: str, newconf: Dict):
+            except Exception as e:
+                continue
+        vhotplugconf = {"vms": vmddplist }
+        os.makedirs(os.path.dirname(self.__ddp_path_wo), exist_ok=True)
```
Everything looks good, except this. I assume that user may also want to manually modify the config. So, please have a look here: [link](https://github.com/tiiuae/FMO-OS/blob/26c20046a41dc872d565636d682bc4701d1ad18b/modules/dynamic-portforwarding-service-host/default.nix#L9)
FMO-tool should be used in service every time it's restarted. And generate config there

buianhhuy96 replied:
Because DDP uses a separate JSON config file, so I believe there should be `__ddp_path_wo`.
And that file is generated from our YAML config file and only based on it. It should only be an internal use for the script.

So what I think we should do is to _hide_ the JSON config, or not allow to manually modify it. If there are conflicts/differences between our YAML and JSON config, it will cause quite alot confusion of which DDP config should be taken into account. 
In the case of manually modifying the YAML config in DDP sections, users can [`fmo-tool ddp generate`](https://github.com/tiiuae/FMO-OS/blob/26c20046a41dc872d565636d682bc4701d1ad18b/modules/dynamic-device-passthrough-services-host/default.nix#L23) to recreate the JSON config based on YAML, but personally I wouldn't suggest that. I only use that command on boot to automatically create the DDP JSON file, then it it updated whenever there is changes in our YAML config file [(whenever there is `save_config()`, there is `write_vmddp_config()` )](https://github.com/tiiuae/fmo-tool/blob/40ca851d4f51af7ecf48939394609c6520b5d549/utils/fmoconfig.py#L76)

jsvapiav replied:
Yep, DPF also has dedicated config. But, this tool supposed just to generate config like: `fmo-tool ddp rules -r > /path/to/json.config`, but not handle this config as is

jsvapiav replied:
okay, I thought a little bit more: that is fine, lets keep this as is. I will rework DPF

jsvapiav commented on utils/fmoconfig.py:
```diff
@@ -21,16 +22,18 @@ class FMOConfig_Manager(object, metaclass=Singleton):
     __valid = False
     __config_path_rw = ""
     __config_path_ro = ""
+    __ddp_path_wo = ""
```
No need to modify config here

jsvapiav commented on apps/devicepassthrough_apps.py:
```diff
@@ -62,8 +74,15 @@ def add_dp_rules(
     dpconfig['devices'] = configuration
+    # If "enable" key not found, set default true
+    if enabled == None:
+        dpconfig['enable'] = True
```
No need to implicitly enable it. May be user wants to add only rules, but not enable it

buianhhuy96 replied:
I think we can set `enable` key to `False` and print a message as below as instruction for user to enable it.
This is because when generate the json config for ddp (see below [`write_vmddp_config`](https://github.com/tiiuae/fmo-tool/blob/40ca851d4f51af7ecf48939394609c6520b5d549/utils/fmoconfig.py#L170) line 170 `fmoconfig.py`), I check for `enable` key. if it is `false`, then it will not be included in generation. So there should be that key, either `False` or `True`, but it should not be missing.

jsvapiav replied:
If it should be then it needs to be set as False by default, not True

jsvapiav replied:
Let's fix this and commit

jsvapiav reviewed (commented):
Look at comments

jsvapiav commented on utils/fmoconfig.py:
```diff
@@ -187,7 +211,8 @@ def _get_vmdci(self, vmname: str) -> FMO_Schema:
     manager = FMOConfig_Manager("/etc/fmo-config.yaml",
-                                "/var/host/fmo-config.yaml")
+                                "/var/host/fmo-config.yaml",
+                                "/var/host/vmddp.conf")
```
At least this needs to be read from `fmo-config.yaml` but let's make a ticket for that

buianhhuy96 committed:
Integrate DDP implementation

- Add new method to write json config for ddp service
- Add command to restart ddp service whenever the config changes

Signed-off-by: Anh Huy Bui <buianhhuy96@gmail.com>
